### PR TITLE
Initial implementation of Cython pickle5 support

### PR DESCRIPTION
**Pull request by suquark**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by suquark** [comment]
The current problem is the buffer copy using `memcpy` could suddenly become 10x slower for larger datasize. Not sure what happened.

**Comment by pcmoritz** [comment]
@suquark Thanks for implementing this, this is great! We can simplify the implementation a bunch: The pickle5 protocol guarantees that all buffers returned by the pickler are contiguous (see https://www.python.org/dev/peps/pep-0574/#buffer-requirements), so we can just put a check in place and remove the code that deals with non-contiguous buffers. This also means we need less metadata (i.e. only the number of the buffers and the sizes of the buffers).

Would it be easier to write the code in C++ with a signature like

```
void WriteBuffersTo(const std::shared_ptr<Buffer>& buffer, const std::vector<LocalMemoryBuffer>& data);
```

where buffer is the buffer that was created by the core worker Create method and data is a vector with all the data buffers from pickle (including the initial bytes)?

**Comment by suquark** [comment]
@pcmoritz  From my understanding, `contiguous` means no `suboffsets`. However, shape, strides and formats may still be necessary even though numpy does not use it. So it think we can just keep the current code.

**Comment by pcmoritz** [comment]
@suquark This looks great! I think we should put the metadata into protobuf, both because it will be easier to understand, and can also reduce the memory (protobuf will omit fields that are not set). I pushed some boilerplate code to do that, can you integrate it with your code? That should be fairly easy given the work you have already done.

There is a slight complication in that we don't know the final offsets of the buffers yet when we construct the metadata, but we can just use relative offsets to the end of the metadata (or the end of the in band buffer). We could also put the metadata as a footer, but then we would need to store the address to that somewhere, so it's probably easier to have it as a header.

You will need to write some more wrappers around the protobuf code, the code can be found in `ray/bazel-genfiles/src/ray/protobuf/serialization.pb.h` after compiling the project :)

File: src/ray/protobuf/serialization.proto [hunk 1]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package ray.serialization;
+
+message PythonBuffer {
```
Comment (pcmoritz) [inline src/ray/protobuf/serialization.proto]: Can you document the fields down here? Especially which buffer location "address" is relative to :)

File: src/ray/protobuf/serialization.proto [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  int32 ndim = 4;
+  bool readonly = 5;
+  string format = 6;
+  repeated int64 shape = 7;
+  repeated int64 strides = 8;
+  // 'suboffsets' is ignored since it is required to be NULL by pickle5 protocol.
+}
+
+message PythonObject {
+  uint64 inband_data_offset = 1;
```
Comment (pcmoritz) [inline src/ray/protobuf/serialization.proto]: Can you also document all of these fields (and specify which buffer locations the offsets are relative to)?

File: python/ray/includes/serialization.pxi [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+            raise ValueError("Total buffer metadata size is bigger than %d. "
+                             "Consider reduce the number of buffers "
+                             "(number of numpy arrays, etc)." % INT32_MAX)
+        self._total_bytes = self._protobuf_offset + protobuf_bytes
+        return self._total_bytes
+
+    cdef void write_to(self, const c_string &inband, shared_ptr[CBuffer] data,
+                       int memcopy_threads):
+        cdef uint8_t *ptr = data.get().Data()
+        cdef uint8_t *__temp
```
Comment (pcmoritz) [inline python/ray/includes/serialization.pxi]: Is this still needed?

File: python/ray/includes/serialization.pxi [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        # Write protobuf size for deserialization.
+        protobuf_size = self.python_object.GetCachedSize()
+        (<int64_t*>ptr)[0] = self._protobuf_offset
+        (<int64_t*>ptr)[1] = protobuf_size
+        # Write protobuf data.
+        self.python_object.SerializeWithCachedSizesToArray(
+            ptr + self._protobuf_offset)
+        # Write inband data.
+        memcpy(ptr + self.python_object.inband_data_offset(),
+               inband.c_str(), inband.length())
```
Comment (pcmoritz) [inline python/ray/includes/serialization.pxi]: Is there a reason you didn't call `inband.data()`? It doesn't need to be null terminated, right?

Comment (suquark) [reply python/ray/includes/serialization.pxi]: oh, that's true. I should call `inband.data()`. thanks for mention it.

File: python/ray/includes/serialization.pxi [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        buffer.len = buffer_meta.length()
+        buffer.itemsize = buffer_meta.itemsize()
+        buffer.readonly = buffer_meta.readonly()
+        buffer.ndim = buffer_meta.ndim()
+        if buffer_meta.format().size() > 0:
+            buffer.format = buffer_meta.release_format().c_str()
+        else:
+            buffer.format = NULL
+        if buffer_meta.shape_size() > 0:
+            buffer.shape = <Py_ssize_t *>malloc(
```
Comment (pcmoritz) [inline python/ray/includes/serialization.pxi]: Is it guaranteed that python will call free on this later?

File: python/ray/includes/serialization.pxi [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            uint64_t inband_data_offset = sizeof(int64_t) * 2
+        buffers_offset = padded_length_u64(
+            inband_data_offset + inband.length(), kMajorBufferAlign)
+        self._protobuf_offset = padded_length_u64(
+            buffers_offset + self._curr_buffer_addr, kMinorBufferAlign)
+        # This is a placeholder for correct calculation of size.
+        self.python_object.set_inband_data_offset(inband_data_offset)
+        self.python_object.set_inband_data_size(inband.length())
+        self.python_object.set_buffers_offset(buffers_offset)
+        # We MUST NOT change 'python_object' size since now.
```
Comment (pcmoritz) [inline python/ray/includes/serialization.pxi]: I don't understand the comment, can you clarify it some more?

**Review by pcmoritz — COMMENTED** [review]
This is great! After doing the fixes can you run the tests with this code path and if it passes we can merge it :)

File: python/ray/includes/serialization.pxi (hunk 5 above)
Comment (suquark) [reply python/ray/includes/serialization.pxi]: yep, I also find out this problem. I need to release it manually.

**Comment by suquark** [comment]
@pcmoritz ready for review again. I have also retested the performance, it shows that protobuf has minor impact.

**Review by pcmoritz — APPROVED** [review]
This looks great! I pushed a small simplification, please have a quick look. I'll re-run the tests on python 3.8 one last time to make sure the changes are good and then we can merge it :)

**Comment by pcmoritz** [comment]
@suquark Even after restoring the number of threads, there is a decrease in memcopy throughput, see the output of the `ray microbenchmark` command:

before:

```
single core get calls per second 19753.07 +- 653.87
single core put calls per second 5043.74 +- 98.92
single core put gigabytes per second 7.51 +- 0.29
multi core put calls per second 7534.58 +- 44.24
multi core put gigabytes per second 3.94 +- 0.43
single core tasks sync per second 1618.65 +- 9.39
single core tasks async per second 3403.5 +- 20.26
multi core tasks async per second 3095.14 +- 32.42
single core actor calls sync per second 1419.07 +- 5.49
single core actor calls async per second 1983.87 +- 23.03
multi core actor calls async per second 2360.99 +- 25.66
```

after:

```
single core get calls per second 29845.42 +- 664.87
single core put calls per second 6462.71 +- 122.62
single core put gigabytes per second 4.25 +- 1.14
multi core put calls per second 9090.25 +- 192.79
multi core put gigabytes per second 3.79 +- 0.29
single core tasks sync per second 1811.24 +- 24.41
single core tasks async per second 3750.24 +- 11.48
multi core tasks async per second 3366.42 +- 31.93
single core actor calls sync per second 1611.36 +- 26.61
single core actor calls async per second 2269.13 +- 18.06
multi core actor calls async per second 2593.69 +- 33.28
```

Do you know what's going on here? We need to fix it before we can turn the flag on by default :)

**Comment by suquark** [comment]
@pcmoritz So except `single core put gigabytes per second 4.25 +- 1.14`, everything gets better, right? I'll check the decrease in memcopy throughput.

**Comment by pcmoritz** [comment]
I am testing on the EC2 instance with python 3.8 installed. Yeah, it's the throughput measurement that I'm concerned about, it does seem to be significantly lower, so it would be good to understand what is going on there. In the worst case we can just use the old code for the multithreaded copy :)

**Pull request merged**